packages/jsapi-utils/src/ViewportDataUtils.ts

```diff
@@ -0,0 +1,158 @@
+import { ListData } from '@react-stately/data';
+import clamp from 'lodash.clamp';
+import { Column, Row, Table, TreeTable } from '@deephaven/jsapi-shim';
+import Log from '@deephaven/log';
+
+export interface KeyedItem<T> {
+  key: string;
+  item?: T;
+}
+
+export type OnTableUpdatedEvent = CustomEvent<{
+  offset: number;
+  rows: ViewportRow[];
+}>;
+
+export type RowDeserializer<T> = (row: ViewportRow, columns: Column[]) => T;
+
+export type ViewportRow = Row & { offsetInSnapshot: number };
+
+const log = Log.module('ViewportDataUtils');
+
+/**
+ * Create a unique string key for a row based on its ordinal position in its
+ * source table. This is calculated based on it's offset in the viewport
+ * (row.offsetInSnapshot) + the offset of the snapshot.
+ * @param row Row from a Table update event.
+ * @param offset Offset of the current viewport.
+ * @returns Unique string key for the oridinal position of the given row.
+ */
+export function createKeyFromOffsetRow(row: ViewportRow, offset: number) {
+  return String(row.offsetInSnapshot + offset);
+}
+
+/**
+ * Creates a handler function for a `dh.Table.EVENT_UPDATED` event. Rows that
+ * get passed to the handler will be added to or updated in the given
+ * `viewportData` object.
+ * @param table DH table instance.
+ * @param viewportData State object for managing a list of KeyedItem data.
+ * @param deserializeRow Converts a DH Row to an item object.
+ * @returns Handler function for a `dh.Table.EVENT_UPDATED` event.
+ */
+export function createOnTableUpdatedHandler<T>(
+  table: Table | TreeTable | null,
+  viewportData: ListData<KeyedItem<T>>,
+  deserializeRow: RowDeserializer<T>
+): (event: OnTableUpdatedEvent) => void {
+  /**
+   * Handler for a `dh.Table.EVENT_UPDATED` event.
+   */
+  return function onTableUpdated(event: OnTableUpdatedEvent) {
+    if (table == null) {
+      return;
+    }
+
+    const { offset, rows } = event.detail;
+
+    log.debug('table updated', event.detail);
+
+    rows.forEach(row => {
+      const item = deserializeRow(row, table.columns);
+
+      const keyedItem = {
+        key: createKeyFromOffsetRow(row, offset),
+        item,
+      };
+
+      if (viewportData.getItem(keyedItem.key) != null) {
+        viewportData.update(keyedItem.key, keyedItem);
+      } else {
+        viewportData.append(keyedItem);
+      }
+    });
+  };
+}
+
+/**
+ * Maps a Row to a key / value object. Keys are mapped ver batim from column
+ * names.
+ * @param row Row to map to an item.
+ * @param columns Columns to map.
+ * @returns A key / value object for the row.
+ */
+export function defaultRowDeserializer<T>(
+  row: ViewportRow,
+  columns: Column[]
+): T {
+  return columns.reduce((result, col) => {
+    // eslint-disable-next-line no-param-reassign
```

Comment:
Wish there was a way to disable this eslint just in reduce functions for that particular parameter.